### PR TITLE
Maintainer and Repository Inactivity Policy

### LINKED ISSUE #32: A default maintainer retirement policy for projects without one
If a project does not have a retirement policy, a default policy of from three to six months of inactivity will result in moving a maintainer to emeritus status

### PR DESCRIPTION
**Pull request by shemnon**
Establish policies for when a maintainer stops contributing and when
repositories stop receiving contributions.

Fixes #32 

Signed-off-by: Danno Ferrin <danno.ferrin@gmail.com>

### CODE REVIEW

File: inactivity.md
Comment (arsulegai) [inline inactivity.md]: A few questions that may arise, just thinking out loud here
1. Repository never needed a maintainer activity.
2. The maintainer is responsible for multiple projects, the maintainer did not have a commit or an activity in one of those projects in the last [3/4/6] months.
3. Should the access be removed from the organization or only from the repository that the user is a maintainer of. If a maintainer is not part of any project then eventually be removed from the org.

File: inactivity.md
Comment (arsulegai) [inline inactivity.md]: There is a separate process for project's lifecycle. We can focus on the maintainers in this section, if we need a revision on project's lifecycle let's do it there.

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]:
```suggestion
As part of the normal lifecycle of a project, codebases and maintainers come and go.  This document formalizes a Hyperledger wide policy for archiving projects and moving maintainers to emeritus status when a project does not have an explicit policy of their own.
```

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]:
```suggestion
Hyperledger very much appreciates the contributions of all maintainers but removing write privileges is in the interest of an orderly and secure project.
```

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]:
```suggestion
Maintainers who have not contributed to a Hyperledger project for (`exact timeframe to be decided`) 3 / 4 / 6 months will be removed from the relevant Github teams.
```

Comment (lehors) [reply inactivity.md]: What counts as a contribution here? I hope this is not limited to PRs and commits. Reviewing and merging PRs is an important part of the maintainers role that we ought to recognize even if that's they have done.

Comment (ryjones) [reply inactivity.md]: @tkuhrt removing them from a team is hard - binning contributions by role is not reliable.

Comment (ryjones) [reply inactivity.md]: @lehors [events that show up in the audit log](https://docs.github.com/en/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/reviewing-the-audit-log-for-your-organization#audit-log-actions) are what we can gate on. There are a lot of activities here - some make sense, some don't.

Comment (tkuhrt) [reply inactivity.md]: > @tkuhrt removing them from a team is hard - binning contributions by role is not reliable.

@ryjones : I was thinking that if a maintainer is part of a project-maintainer team, then we would want to remove them from that. Are you saying that the preferred mechanism would be to drop them completely from the organization? What does that do for GitHub teams they were previously part of? Does it automatically remove them from those teams?

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]:
```suggestion
Activity can be code contributions, code reviews, issue reporting, or any other such activity trackable by GitHub attributed to a Hyperledger repository.
```

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]:
```suggestion
Maintainers will be notified of this via an "at" @ mention in a pull request moving their name from a projects active contributors to the emeritus contributors.
```

File: inactivity.md
Comment (ryjones) [reply inactivity.md]: @tkuhrt you're right, I conflated two discussions :)

**Comment by ryjones** [comment]
@lehors [here is the list of actions committed this year, and the count, by non-robots and non-LF-employees](https://gist.github.com/ryjones/816444675103a234dfff553d40200bf2)

File: inactivity.md
Comment (C0rWin) [inline inactivity.md]:
```suggestion
Maintainers who express a desire to continue contributing will be given a three-month extension to contribute to the project at the maintainer's request.
```

File: inactivity.md
Comment (C0rWin) [inline inactivity.md]:
```suggestion
These policies apply to projects that do not have an explicit maintainer inactivity policy. Where a project has an established and functioning policy, the project's policies shall prevail.
```

File: inactivity.md
Comment (C0rWin) [inline inactivity.md]:
```suggestion
Projects will be notified via a PR in the appropriate repository and a notice in the project corresponding Discord channel and mailing list if they exist.
```

File: inactivity.md
Comment (C0rWin) [inline inactivity.md]:
```suggestion
Generally speaking, projects that expressed a request to keep the repositor active will be honoured. However, this request may not be honoured if a project has many out of date dependencies, particularly ones relating to security vulnerabilities.
```

File: inactivity.md
Comment (C0rWin) [inline inactivity.md]:
```suggestion
A request by a project to un-archive a repository for active contribution will be honoured unless the project is in a dormant or end-of-life state. In those cases, the project lifecycle issues will need to be resolved first.
```

File: inactivity.md
Comment (C0rWin) [inline inactivity.md]: I think we need to reflect somehow that commits to being counted has to be of significant importance and not small updates and changes. Another point is that there are repos where maintainers extracted self-contained functionality and there is not actually too much to update there besides probably some dependencies or release notes. IMO we also need to be careful with this here, I would consider per project health status rather than the single repo.

Comment (shemnon) [reply inactivity.md]: For a HLF wide policy we need a very objective policy, and judging the worth of a commit is about as subjective as it gets.  Individual projects can make their own policy if they want to judge the significance of a contribution.

Regarding the static "framework" repositories, the clause about requests not to close has no limit (unlike the contributor policy) so a true framework can be kept unarchived indefinitely, until the security notifications start piling up.

One reason to focus this per-repository instead of per-project is that some project repos have been forgotten by their projects when maintainers leave. If the repo is valuable the team can request it not be archived.

File: inactivity.md
Comment (shemnon) [reply inactivity.md]: 1. ?? I'm not sure what is being asked ??
2. The tooling appears to allow us to track per-repo, so we can enforce this on a per-project basis. i.e. if I am part of project Foo and project Bar and stop contributing to Bar for 6 months, my maintainer status on Bar is removed and my commit access for Bar is removed.  My status in Foo is unaffected
3. as discussed further down the list, it will be per team. If they are removed from all teams then org removal would be the next step.

File: inactivity.md
Comment (shemnon) [reply inactivity.md]: I would prefer to keep all inactivity policies in one place.  Splitting it into two policies splits the focus, which is dealing with inactivity holistically and not piecemeal.

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]: Since we have a project lifecycle, we would not be archiving projects with this policy, but rather repositories.
```suggestion
As part of the normal lifecycle of a project codebases and maintainers come and go. This document formalizes a Hyperledger wide policy for archiving repositories and moving maintainers to emeritus status when a project does not have an explicit policy of their own.
```

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]: Best to clarify the inactive maintainer vs. the active maintainers.
```suggestion
The inactive maintainer will be notified of this via an "at" @ mention in a pull request moving their name from a project's active maintainers to the emeritus maintainers.
```

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]: Not sure if the second maintainer in this statement refers to the active maintainers or the identified inactive maintainer. It would be good to clarify that the identified inactive maintainer is the one making the request. Do the active maintainers have the ability to reject this request? Also, is there a limit to the number of times that the identified inactive maintainer can make that request?

File: inactivity.md
Comment (tkuhrt) [reply inactivity.md]: It is possible that repositories for an otherwise active project have become inactive. It is important to have some mechanism to "archive" those inactive repositories. I agree that any project lifecycle changes should be in the project lifecycle document.

File: inactivity.md
Comment (shemnon) [reply inactivity.md]: I will workshop the phrasing. The intent is that the maintainer who is proposed to be removed can make the request, and they get a single three month extension.  If they need more than 9 months my view off they should go emeritus and when they are ready to return re-ask for active status with the current maintainers.

Comment (tkuhrt) [reply inactivity.md]: Agree with a single extension. Just want that to be obvious so there are no chances of someone who just keeps asking for extensions and saying that it is okay based on the policy.

File: inactivity.md
Comment (denyeart) [inline inactivity.md]: 3 months sounds too short.
6 months + 3 month extension sounds too long (9 months total).
My suggestion would be to split the difference: a 4 month inactivity timeframe, with option for 2 month extension, resulting in a total period of 6 months.

**Review by tkuhrt — APPROVED** [review]

**Comment by tkuhrt** [comment]
I vote for 3 months of inactivity with one 3 month extension granted upon request.

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]:
```suggestion
As part of the normal lifecycle of a project codebases and maintainers come and go. This document formalizes a Hyperledger wide policy for moving maintainers to emeritus status when a project does not have an explicit policy of their own
```

at least until we re-add the repository inactivity portion.

File: index.md
Comment (tkuhrt) [inline index.md:17]:
```suggestion
* [Maintainer Inactivity](./inactivity.md)
```

At least until we re-add the repository section.

**Review by petermetz — APPROVED** [review]

File: inactivity.md
Comment (tkuhrt) [inline inactivity.md]: Who will open the proposal?

File: index.md
Comment (tkuhrt) [reply index.md:17]: @shemnon : Why was this marked as resolved?

File: inactivity.md
Comment (shemnon) [reply inactivity.md]: Anyone from the TSC?  Or the TSC chair or who the TSC chair designates?

Comment (denyeart) [reply inactivity.md]: I'm good with the overall proposal here and I am approving. Do we need to indicate that a majority of project maintainers should approve the request prior to merging? Or is it assumed that each project defines such policies?

(Side note: getting a majority of maintainers to chime in on a PR can be difficult, especially if there are inactive maintainers - this is one of the reasons why it helps to retire inactive maintainers. In general it is better to build consensus than have votes by majority rule, but for some contentious decisions such as retirements having a majority vote on record helps to finalize a decision).

**Review by denyeart — APPROVED** [review]

File: inactivity.md
Comment (shemnon) [reply inactivity.md]: This is a default policy that does not involve maintainer action, simply maintainer inaction.  Hence if a maintainer ceases to be active and a PR is posted and no other maintainers approve the PR it will still merge. It doesn't need a majority of the maintainers or even a majority of those voicing an opinion.

Being a default policy if the maintainers don't like it they should update their own policy, so long as a policy exists and is functional this policy has no impact (assuming this policy passes).

**Comment by tkuhrt** [comment]
@hyperledger/tsc : Just a reminder to approve or comment on this PR. We do not currently have enough votes here in GitHub to approve. If we do not get that, we will bring this up for a vote in Thursday's meeting.

**Review by lehors — APPROVED** [review]
LGTM. Thanks!

**Review by C0rWin — APPROVED** [review]

**Review by ryjones — APPROVED** [review]

**Comment by shemnon** [comment]
I vote YES.  I cannot approve my own PR so this will be my vote. 

Posting a PR may be taken as an implicit YES vote, but I can see situations where making a proposal to vote negatively on would be useful to crystalize committee policy and positions.

**Comment by knagware9** [comment]
yes

File: inactivity.md
Comment (arsulegai) [reply inactivity.md]: For item (1) I was referring to a highly possible case, let's consider a project, where one of the maintainer is responsible only to maintain a specific repository. But it happened so that the repository is pretty much stable and did not require any update for months.

File: inactivity.md
Comment (arsulegai) [inline inactivity.md:20]: Does this mean that TSC chair can choose their preferred media of communication to different projects and repositories?
I wouldn't think that's necessary, moving to emeritus status is anyway publicly visible. There can be a standard process setup where notifications are sent to a group of maintainers. For example, a discord channel for maintainers of that project, on mailing list etc. The secondary means of communication can then become optional.

**Review by arsulegai — APPROVED** [review]

File: inactivity.md
Comment (shemnon) [reply inactivity.md:20]: I don't want to write the means of warning in to the policy.  This makes it flexible if a project changes chat servers again, or if one project uses email more than chat, or any other such eventualities.

Since this is a courtesy notice it may not come in the form of a PR.

**Comment by ryjones** [comment]
[Approved by voice vote](https://wiki.hyperledger.org/display/TSC/2022+05+12+TSC+Meeting+Record) 12 MAY 2022

**Pull request merged**
